### PR TITLE
allow cancelation exceptions through the PE stream diagnostic filter

### PR DESCRIPTION
**Pull request by TyOverby**

### CODE REVIEW

**Comment by TyOverby** [comment]
@jaredpar @davkean @jasonmalinowski

**Comment by jaredpar** [comment]
@MattGertz seeking ask mode approval.  

The earlier change to catch exceptions here during emit I/O failed to add `OperationCancelledException` to the exclusion list.  That means that cancelling emit changed from throwing an exception to returning fails and emitting a diagnostic.  There is strong belief this is at the root of the cross language flakiness.

**Comment by jaredpar** [comment]
CC @dotnet/roslyn-compiler

**Comment by jaredpar** [comment]
LGTM

**Comment by MattGertz** [comment]
Approved pending successful test completion.

**Comment by jaredpar** [comment]
CC @CyrusNajmabadi

**Comment by davkean** [comment]
:+1:

**Comment by CyrusNajmabadi** [comment]
:+1: 

Though i'm concerned about any deeper try/catches that may be catching too much.

**Comment by CyrusNajmabadi** [comment]
N/m  ty/jared gave good info why not to be concerned.  (namely, this exception is thrown deeeeeeep within emitting, and the test verified it makes it all the way out to the caller).

**Comment by davkean** [comment]
Let's get this one in, even though this branched is closed - this is a fix for what believe to be the cause of the last major set of flakiness.

**Comment by davkean** [comment]
(Once tests passed of course)

File: src/Test/Utilities/Shared/Pe/BrokenStream.cs [hunk 1]
```diff
@@ -68,7 +68,13 @@ public override void SetLength(long value)
         public override void Write(byte[] buffer, int offset, int count)
         {
             if (BreakHow == 0)
+            {
                 throw new IOException();
+            }
+            if (BreakHow == 3)
```
Comment (jasonmalinowski) [inline src/Test/Utilities/Shared/Pe/BrokenStream.cs]: else if? It's equivalent but might read easier.

**Comment by jasonmalinowski** [comment]
:+1:

File: src/Test/Utilities/Shared/Pe/BrokenStream.cs (hunk 1 above)
Comment (TyOverby) [inline src/Test/Utilities/Shared/Pe/BrokenStream.cs]: I'll incorporate this when I solve your merge conflicts when flowing this into master.  (I don't want to have to run the test suite again right now...)

**Comment by davkean** [comment]
You just beat me. :)

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs [hunk 2]
```diff
@@ -2976,5 +2976,24 @@ public static void Main()
...
+public class X
+{
+    public static void Main()
+    {
+  
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib(source);
+            var broken = new BrokenStream();
+            broken.BreakHow = 3;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs]: @TyOverby I think we should switch to using an enum values with descriptive names.

**Pull request merged**